### PR TITLE
Alternate `ViewStore.binding` helper

### PR DESCRIPTION
**Pull request by stephencelis**
It occurred to me we could just leverage dynamic member lookup and derive view store bindings given simply a binding action...there's a caveat, though.

### CODE REVIEW

File: Sources/ComposableArchitecture/SwiftUI/Binding.swift [hunk 1]
```diff
@@ -258,4 +258,29 @@ extension ViewStore {
...
+  /// TextField(
+  ///   "Enter text",
+  ///   text: viewStore.binding(Action.binding).text
+  /// )
+  /// ```
+  ///
+  /// - Parameter action: A function that wraps a binding action in the view store's action type.
+  /// - Returns: A binding.
+  public func binding(_ action: @escaping (BindingAction<State>) -> Action) -> Binding<State>
+  where State: Equatable {
```
Comment (stephencelis) [inline Sources/ComposableArchitecture/SwiftUI/Binding.swift:280]: All of state needs to be equatable to return a `Binding` here. This is a common enough practice in TCA apps, but an obstacle nonetheless.

We could overcome that by having a `ViewStore.Binding` helper that has dynamic member lookup over equatable sub-state, instead, but wanted to start a convo first to see if it's worth introducing that type just to clean up call sites a lil bit.

File: Examples/CaseStudies/SwiftUICaseStudies/01-GettingStarted-Bindings-Forms.swift [hunk 2]
```diff
@@ -56,20 +56,20 @@ struct BindingFormView: View {
           HStack {
             TextField(
               "Type here",
-              text: viewStore.binding(keyPath: \.text, send: BindingFormAction.binding)
+              text: viewStore.binding(BindingFormAction.binding).text
```
Comment (stephencelis) [inline Examples/CaseStudies/SwiftUICaseStudies/01-GettingStarted-Bindings-Forms.swift:59]: It's decently shorter, so might be worth it. It also reads a bit jumbly right now: a lot of `binding.binding` noise. I wonder if we can come up with a less repetitive call site.

File: Sources/ComposableArchitecture/SwiftUI/Binding.swift (hunk 1 above)
Comment (mbrandonw) [reply Sources/ComposableArchitecture/SwiftUI/Binding.swift:280]: This is already the case with the current `.binding(keyPath:send:)` right?

File: Examples/CaseStudies/SwiftUICaseStudies/01-GettingStarted-Bindings-Forms.swift (hunk 2 above)
Comment (mbrandonw) [reply Examples/CaseStudies/SwiftUICaseStudies/01-GettingStarted-Bindings-Forms.swift:59]: yeah I like this!

File: Sources/ComposableArchitecture/SwiftUI/Binding.swift (hunk 1 above)
Comment (stephencelis) [reply Sources/ComposableArchitecture/SwiftUI/Binding.swift:280]: Nah that's only equatable on the `Value`. This requires equatable on `Root`.

**Comment by iampatbrown** [comment]
Nice addition. I've always played around with the binding call site and never been able to come up with something that feels right. Keen to see where this leads :)

I tried something like this at one point but didn't really like the idea of using a protocol.
```swift
public protocol Bindable {
  associatedtype State
  static func binding(_: BindingAction<State>) -> Self
}

extension ViewStore where State: Equatable, Action: Bindable, Action.State == State {
  public var binding: Binding<State> { self.binding(Action.binding) }
}

extension Reducer where Action: Bindable, Action.State == State {
  public var binding: Self {  self.binding(action: /Action.binding) }
}
```
leading to 
```swift
viewStore.binding.text
```

**Comment by stephencelis** [comment]
@iampatbrown Yup! @junebash brought that up that great idea in https://github.com/pointfreeco/swift-composable-architecture/discussions/370

I'm going to close this draft for now as we kick around a few more ideas, though :smile:

**Pull request closed without merging**